### PR TITLE
feat(mcp): expand to 12 tools + Streamable HTTP transport

### PR DESCRIPTION
**Pull request by xang1234**
## Summary

- **4 new MCP tools**: `group_rankings`, `stock_lookup`, `breadth_snapshot`, `daily_digest` — expanding coverage from 8 to 12 tools
- **Streamable HTTP transport**: MCP server now reachable at `POST /mcp` (JSON-RPC over HTTP) alongside existing stdio, gated by `MCP_HTTP_ENABLED` (default: true)
- **Quality hardening**: consolidated DB sessions, thread-safe lazy singleton with `run_in_threadpool`, public `dispatch()` API on `MarketCopilotMcpServer`, fixed silent date-type bugs in `_fact`/`_freshness` helpers

## New tools

| Tool | Description | Service/Source |
|------|-------------|----------------|
| `group_rankings` | IBD industry group rankings + rank movers | `IBDGroupRankService` |
| `stock_lookup` | Stock universe info + optional technicals | `StockUniverse` + feature store |
| `breadth_snapshot` | N-day market breadth history | `MarketBreadth` |
| `daily_digest` | Full daily digest (stance, leaders, themes, risks) | `DigestService` |

## HTTP transport

- `POST /mcp/` — JSON-RPC request/response
- `GET /mcp/` — 405 (SSE placeholder)
- `DELETE /mcp/` — session teardown (no-op)
- Sync DB work offloaded via `run_in_threadpool` to avoid blocking the event loop
- Lazy singleton with double-checked locking (`threading.Lock`)

## Test plan

- [x] 20 unit tests passing (`test_mcp_market_copilot.py`) — tool count, all 4 new handlers, edge cases
- [x] 8 integration tests passing (`test_mcp_server_integration.py`) — stdio + 7 HTTP transport tests
- [x] Full unit suite (374/375 pass, 1 pre-existing flaky test unrelated)
- [ ] Manual smoke test: `curl -X POST http://localhost:8000/mcp/ -H 'Content-Type: application/json' -d '{"jsonrpc":"2.0","id":1,"method":"tools/list","params":{}}'`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/xang1234/stock-screener/pull/25" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

## Summary by CodeRabbit

* **New Features**
  * Configurable Streamable HTTP MCP transport (can be enabled/disabled).
  * Four new MCP tools: group rankings, stock lookup, market breadth snapshot, and daily digest.
  * Docs updated: MCP capabilities now list 12 tools and mention stdio + Streamable HTTP transports.

* **Chores**
  * Expanded test coverage for the HTTP transport and the new MCP tools.

### CODE REVIEW

File: backend/tests/integration/test_mcp_server_integration.py [hunk 1]
```diff
@@ -91,3 +94,124 @@ def test_mcp_server_lists_tools_and_serves_market_overview(tmp_path):
...
+@pytest.mark.asyncio
+class TestMcpHttpTransport:
+    """Integration tests for the Streamable HTTP MCP transport."""
+
+    @pytest_asyncio.fixture()
+    async def client(self, tmp_path):
+        """Create an httpx client backed by the FastAPI app with seeded test data."""
+        database_path = tmp_path / "mcp-http.sqlite3"
+        database_url = f"sqlite:///{database_path}"
+
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/tests/integration/test_mcp_server_integration.py:117]: Not applicable — all MCP unit/integration tests intentionally use isolated temporary SQLite databases (`tmp_path`) to avoid corrupting the production `data/stockscanner.db`. The CLAUDE.md rule about database location applies to **application runtime**, not test fixtures. The existing stdio integration test in this same file already follows the same pattern (`tmp_path / "mcp-server.sqlite3"`), and the project's `mcp_fixture.py` helper is specifically designed for ephemeral in-memory/file databases.

File: backend/app/interfaces/mcp/http_transport.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                "id": None,
+                "error": {"code": -32700, "message": "Parse error: invalid JSON"},
+            },
+            status_code=200,
+            media_type="application/json",
+        )
+
+    server = _get_server()
+    response = await run_in_threadpool(server.dispatch, body)
+
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/interfaces/mcp/http_transport.py:76]: Fixed in 47c2400. Added:
1. Body validation in `http_transport.py` — non-dict JSON (arrays, scalars) now returns `-32600 Invalid Request` instead of 500
2. `dispatch()` in `server.py` hardened with `isinstance(message, dict)` guard + catch-all `except Exception` → `-32603 Internal error`
3. New test `test_non_dict_body_returns_invalid_request` verifies `[1,2,3]` returns -32600

File: backend/app/interfaces/mcp/http_transport.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                "jsonrpc": "2.0",
+                "id": None,
+                "error": {"code": -32700, "message": "Parse error: invalid JSON"},
+            },
+            status_code=200,
+            media_type="application/json",
+        )
+
+    server = _get_server()
+    response = await run_in_threadpool(server.dispatch, body)
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/interfaces/mcp/http_transport.py:75]: Fixed in 47c2400. Added `isinstance(body, dict)` check in `http_transport.py` (returns -32600) + defense-in-depth in `dispatch()` (returns -32603 on unexpected exceptions). New test covers the array case.

File: backend/app/interfaces/mcp/market_copilot.py [hunk 4]
```diff
@@ -797,6 +825,210 @@ def _watchlist_add(self, args: WatchlistAddArgs) -> ToolEnvelope:
...
+        ranking_date = date.fromisoformat(ranking_date_str) if ranking_date_str else None
+        top_rankings = [
+            {
+                "industry_group": r["industry_group"],
+                "rank": r["rank"],
+                "avg_rs_rating": r.get("avg_rs_rating"),
+                "num_stocks": r.get("num_stocks"),
+                "top_symbol": r.get("top_symbol"),
+                "rank_change_1w": r.get("rank_change_1w"),
+            }
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/interfaces/mcp/market_copilot.py:858]: Fixed in 47c2400. Replaced hardcoded `rank_change_1w` with dynamic `rank_change_key = f"rank_change_{args.period}"`. The output field is now `rank_change` (period-agnostic) and reads from the correct period-specific key in the service response.

File: backend/app/interfaces/mcp/models.py [hunk 5]
```diff
@@ -150,3 +150,34 @@ def normalize_symbols(cls, value: list[str]) -> list[str]:
...
+class StockLookupArgs(BaseModel):
+    """Arguments for stock_lookup."""
+
+    symbol: str = Field(..., min_length=1, max_length=10)
+    include_technicals: bool = False
+
+    @field_validator("symbol")
+    @classmethod
+    def normalize_symbol(cls, value: str) -> str:
+        return value.strip().upper()
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/interfaces/mcp/models.py]: Fixed in 47c2400. `StockLookupArgs.normalize_symbol` now raises `ValueError` when the stripped value is empty — `"   "` correctly fails Pydantic validation.

File: backend/app/interfaces/mcp/server.py [hunk 6]
```diff
@@ -93,6 +93,10 @@ def serve_forever(self) -> None:
             if response is not None:
                 self._transport.write_message(response)
 
+    def dispatch(self, message: dict[str, Any]) -> dict[str, Any] | None:
+        """Dispatch a JSON-RPC message and return the response (or None for notifications)."""
+        return self._handle_message(message)
+
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/interfaces/mcp/server.py:104]: Fixed in 47c2400. `dispatch()` now validates `isinstance(message, dict)` (returns -32600) and wraps `_handle_message` in try/except (returns -32603). Two layers of defense: HTTP transport rejects non-dicts before they reach dispatch, and dispatch catches anything else.

File: backend/tests/integration/test_mcp_server_integration.py [hunk 7]
```diff
@@ -91,3 +94,124 @@ def test_mcp_server_lists_tools_and_serves_market_overview(tmp_path):
...
+        test_server = MarketCopilotMcpServer(test_service, transport=None)
+
+        # Override the lazy singleton
+        original_server = http_mod._server
+        http_mod._server = test_server
+
+        from app.main import app
+
+        transport = httpx.ASGITransport(app=app)
+        async with httpx.AsyncClient(transport=transport, base_url="http://test") as client:
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/tests/integration/test_mcp_server_integration.py]: Fixed in 47c2400. Test fixture now builds an isolated `FastAPI()` app with only the MCP router mounted, completely decoupled from `app.main` and the import-time `MCP_HTTP_ENABLED` gate. No more dependency on process env or prior imports.

File: backend/app/main.py [hunk 8]
```diff
@@ -493,6 +494,11 @@ async def health_check():
 from .api.v1.router import router as api_router
 app.include_router(api_router, prefix="/api/v1")
 
+# Mount MCP Streamable HTTP transport (JSON-RPC over HTTP)
+if settings.mcp_http_enabled:
+    from .interfaces.mcp import mcp_router
+    app.include_router(mcp_router, prefix="/mcp")
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/main.py]: Fixed in 8c5b8f1. The `/mcp` router is now mounted with `dependencies=[Depends(require_server_session)]`, matching exactly how the `/api/v1` routes are protected. When `SERVER_AUTH_ENABLED=true`, all MCP endpoints require a valid session cookie or bearer token. When auth is disabled (local dev), the dependency is a no-op.

File: backend/app/interfaces/mcp/market_copilot.py [hunk 9]
```diff
@@ -797,6 +825,211 @@ def _watchlist_add(self, args: WatchlistAddArgs) -> ToolEnvelope:
...
+            for r in rankings
+        ]
+
+        return self._envelope(
+            f"Top {len(top_rankings)} IBD group rankings as of {ranking_date}. "
+            f"Showing {args.period} movers ({len(movers.get('gainers', []))} gainers, {len(movers.get('losers', []))} losers).",
+            facts=[
+                self._fact("ranking_date", str(ranking_date), "ibd_group_ranks"),
+                self._fact("total_groups", len(rankings), "ibd_group_ranks"),
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/app/interfaces/mcp/market_copilot.py]: Fixed in 8c5b8f1. Renamed `total_groups` → `returned_groups` to accurately reflect the fact value is bounded by the caller's `limit` parameter.

File: backend/tests/integration/test_mcp_server_integration.py [hunk 10]
```diff
@@ -91,3 +94,133 @@ def test_mcp_server_lists_tools_and_serves_market_overview(tmp_path):
...
+        test_settings = SimpleNamespace(
+            mcp_server_name="test-mcp-http",
+            mcp_watchlist_writes_enabled=False,
+        )
+        test_service = MarketCopilotService(session_factory, test_settings)
+        test_server = MarketCopilotMcpServer(test_service, transport=None)
+
+        # Override the lazy singleton so the router uses our test server
+        original_server = http_mod._server
+        http_mod._server = test_server
```
[the remark this answers is not in the document]
Comment (xang1234) [reply backend/tests/integration/test_mcp_server_integration.py:130]: Fixed in 8c5b8f1. The fixture now uses `try/finally` to guarantee `http_mod._server` is restored and `test_engine.dispose()` is called even if fixture setup or the test itself fails.
